src/codegen/compvars.cpp

```diff
@@ -74,43 +74,62 @@ std::string ValuedCompilerType<llvm::Value*>::debugName() {
     return rtn;
 }
 
+// Returns a CompilerVariable representing the given variable's type
+static CompilerVariable* getClassCompilerVariable(IREmitter& emitter, ConcreteCompilerVariable* var) {
+    // val->cls
+    static_assert(offsetof(Box, cls) == 0, "");
+    llvm::Value* cls_ptr = emitter.getBuilder()->CreateConstInBoundsGEP2_32(var->getValue(), 0, 0);
+    llvm::Value* cls = emitter.getBuilder()->CreateLoad(cls_ptr);
+    assert(cls->getType() == g.llvm_class_type_ptr);
+    // cast from BoxedClass* to Box*
+    llvm::Value* cls_as_box = emitter.getBuilder()->CreateGEP(
+        cls, { llvm::ConstantInt::get(g.i32, 0), llvm::ConstantInt::get(g.i32, 0), llvm::ConstantInt::get(g.i32, 0) });
+    assert(cls_as_box->getType() == g.llvm_value_type_ptr);
+    return new ConcreteCompilerVariable(UNKNOWN, cls_as_box, false);
+}
+
 struct RawInstanceMethod {
-    CompilerVariable* obj, *func;
+    CompilerVariable* obj, *func, *im_class;
 
-    RawInstanceMethod(CompilerVariable* obj, CompilerVariable* func) : obj(obj), func(func) {}
+    RawInstanceMethod(CompilerVariable* obj, CompilerVariable* func, CompilerVariable* im_class)
+        : obj(obj), func(func), im_class(im_class) {}
 };
 
 class InstanceMethodType : public ValuedCompilerType<RawInstanceMethod*> {
 private:
-    static std::unordered_map<std::pair<CompilerType*, CompilerType*>, InstanceMethodType*> made;
+    static std::unordered_map<std::pair<CompilerType*, std::pair<CompilerType*, CompilerType*>>, InstanceMethodType*>
+        made;
 
-    CompilerType* obj_type, *function_type;
-    InstanceMethodType(CompilerType* obj_type, CompilerType* function_type)
-        : obj_type(obj_type), function_type(function_type) {}
+    CompilerType* obj_type, *function_type, *class_type;
+    InstanceMethodType(CompilerType* obj_type, CompilerType* function_type, CompilerType* class_type)
+        : obj_type(obj_type), function_type(function_type), class_type(class_type) {}
```

Comment:
It looks like class_type is unused -- we include the other two types here since they are needed for type analysis, but it seems like we don't need to add class_type?  ie the type behavior is already determined by the pair (obj_type, function_type).